stgit/commands/rebase.py

```diff
@@ -61,11 +77,153 @@
 
 directory = DirectoryGotoTopLevel()
 
+INTERACTIVE_APPLY_LINE = '# --- APPLY_LINE ---'
+
+INTERACTIVE_HELP_INSTRUCTIONS = """
+# Commands:
+# k, keep <patch> = do not modify this patch
+# s, squash <patch> = squash patch into the previous patch
+# d, delete <patch> = delete patch
+#
+# These lines can be re-ordered; they are executed from top to bottom.
+#
+# Patches above the APPLY_LINE are applied; other patches are kept unapplied.
+"""
+
+
+def __get_description(stack, patch):
+    """Extract and return a patch's short description"""
+    cd = stack.patches[patch].data
+    return cd.message_str.strip().split('\n', 1)[0].rstrip()
+
+
+class Action(Enum):
+    # do nothing; a 'no-op'
+    KEEP = 0
+    # delete the patch
+    DELETE = 1
+    # squash the patch into the previous patch
+    SQUASH = 2
+
+
+Instruction = typing.NamedTuple(
+    'Instruction', [('patch_name', str), ('action', Action), ('apply', bool)]
+)
+
+
+def __do_rebase_interactive(repository, previously_applied_patches, check_merged):
+    """Opens an interactive editor, generates instruction list, and executes instructions."""
+    stack = repository.get_stack()
+
+    if len(stack.patchorder.all) == 0:
+        return utils.STGIT_SUCCESS
+
+    name_len = max((len(s) for s in stack.patchorder.all))
+    line = 'keep {{:<{name_len}}} # {{}}'.format(name_len=name_len)
+
+    # create a list of all patches to send to the editor
+    instructions = []
+    for pn in previously_applied_patches:
+        patch = (pn, __get_description(stack, pn))
+        instructions.append(line.format(*patch))
+    instructions.append(INTERACTIVE_APPLY_LINE)
+    for pn in stack.patchorder.all:
+        if pn in previously_applied_patches:
+            continue
+        patch = (pn, __get_description(stack, pn))
+        instructions.append(line.format(*patch))
+    instructions.append(INTERACTIVE_HELP_INSTRUCTIONS)
+
+    # open an editor to let the user generate the 'todo' instructions
+    todo = edit_string(
+        '\n'.join(instructions),
+        '.stgit-rebase-interactive.txt',
+    )
+
+    # parse the instructions we've been given
+    seen_apply_line = False
+    instructions = []
+    for line in todo.split('\n'):
+        line = line.strip()
+        if line == INTERACTIVE_APPLY_LINE:
+            seen_apply_line = True
+
+        # ignore comment lines
+        if not line or line.startswith('#'):
+            continue
```

Comment:
A keyboarding mistake I sometimes make is to type `J` in vim, which concatenates the subsequent line with the current line. I'm thinking about a case where the user's instructions end up like this:

```
keep p0
# --- APPLY LINE ---keep p1
keep p2
```
So I'm wondering if it's worth the trouble of searching for the apply line more robustly such that we can warn or error-out if the user puts junk after (or before) the apply line string?

I'm not sure anything must be done about this; just want you to consider it.

Reply:
Fixed! Added a test for it.

```
        # record when we find the APPLY_LINE so we know which patches to apply
        if INTERACTIVE_APPLY_LINE in line:
            if INTERACTIVE_APPLY_LINE == line:
                seen_apply_line = True
            else:
                raise CmdException("Bad APPLY_LINE: '%s'" % line)
``